Loop/Managers/Missed Meal Detection/MealDetectionManager.swift

```diff
@@ -0,0 +1,311 @@
+                effectValueCache[effect.startDate] = (effectValueCache[effect.startDate] ?? 0) + value
+            }
+            
+            var unexpectedDeviation: Double = 0
+            var mealTime = now
+            
+            /// Dates the algorithm uses when computing effects
+            /// Have the range go from newest -> oldest time
+            let summationRange = LoopMath.simulationDateRange(from: intervalStart,
+                                                    to: now,
+                                                    delta: delta)
+                                          .reversed()
+            
+            /// Dates the algorithm is allowed to check for the presence of a missed meal
+            let dateSearchRange = Set(LoopMath.simulationDateRange(from: intervalStart,
+                                                         to: intervalEnd,
+                                                         delta: delta))
+            
+            /// Timeline used for debug purposes
+            var missedMealTimeline: [(date: Date, unexpectedDeviation: Double?, mealThreshold: Double?, rateOfChangeThreshold: Double?)] = []
+            
+            for pastTime in summationRange {
+                guard
+                    let unexpectedEffect = effectValueCache[pastTime],
+                    !carbEntries.contains(where: { $0.startDate >= pastTime })
+                else {
+                    missedMealTimeline.append((pastTime, nil, nil, nil))
+                    continue
+                }
+                
+                unexpectedDeviation += unexpectedEffect
+
+                guard dateSearchRange.contains(pastTime) else {
+                    /// This time is too recent to check for a missed meal
+                    missedMealTimeline.append((pastTime, unexpectedDeviation, nil, nil))
+                    continue
+                }
+                
+                /// Find the threshold based on a minimum of `missedMealGlucoseRiseThreshold` of change per minute
+                let minutesAgo = now.timeIntervalSince(pastTime).minutes
+                let deviationChangeThreshold = MissedMealSettings.glucoseRiseThreshold * minutesAgo
+                
+                /// Find the total effect we'd expect to see for a meal with `carbThreshold`-worth of carbs that started at `pastTime`
+                guard let modeledMealEffectThreshold = self.effectThreshold(mealStart: pastTime, carbsInGrams: MissedMealSettings.minCarbThreshold) else {
+                    continue
+                }
+                
+                missedMealTimeline.append((pastTime, unexpectedDeviation, modeledMealEffectThreshold, deviationChangeThreshold))
+                
+                /// Use the higher of the 2 thresholds to ensure noisy CGM data doesn't cause false-positives for more recent times
+                let effectThreshold = max(deviationChangeThreshold, modeledMealEffectThreshold)
```

Comment:
It was a little difficult for me to understand why there were two thresholds. It looks like one is rate based (deviationChangeThreshold) and the other is meal size based (modeledMealEffectThreshold).  Maybe a naming change could make this more clear?